src/test/fuzz/miniscript.cpp

```diff
@@ -0,0 +1,167 @@
+ * to focus on fuzzing the Miniscript nodes' test representation, not the key representation.
+ */
+struct ParserContext {
+    typedef CPubKey Key;
+
+    bool KeyCompare(const Key& a, const Key& b) const {
+        return a < b;
+    }
+
+    std::optional<std::string> ToString(const Key& key) const
+    {
+        auto it = TEST_DATA.dummy_key_idx_map.find(key);
+        if (it == TEST_DATA.dummy_key_idx_map.end()) return {};
+        uint8_t idx = it->second;
+        return HexStr(Span{&idx, 1});
+    }
+
+    template<typename I>
+    std::optional<Key> FromString(I first, I last) const {
+        if (last - first != 2) return {};
+        auto idx = ParseHex(std::string(first, last));
+        if (idx.size() != 1) return {};
+        return TEST_DATA.dummy_keys[idx[0]];
+    }
+
+    template<typename I>
+    std::optional<Key> FromPKBytes(I first, I last) const {
+        Key key;
+        key.Set(first, last);
+        if (!key.IsValid()) return {};
+        return key;
+    }
+
+    template<typename I>
+    std::optional<Key> FromPKHBytes(I first, I last) const {
+        assert(last - first == 20);
+        CKeyID keyid;
+        std::copy(first, last, keyid.begin());
+        const auto it = TEST_DATA.dummy_keys_map.find(keyid);
+        if (it == TEST_DATA.dummy_keys_map.end()) return {};
+        return it->second;
+    }
+} PARSER_CTX;
+
+//! Context that implements naive conversion from/to script only, for roundtrip testing.
+struct ScriptParserContext {
+    //! For Script roundtrip we never need the key from a key hash.
+    struct Key {
+        bool is_hash;
+        std::vector<unsigned char> data;
+    };
+
+    bool KeyCompare(const Key& a, const Key& b) const {
+        return a.data < b.data;
+    }
+
+    const std::vector<unsigned char>& ToPKBytes(const Key& key) const
+    {
+        assert(!key.is_hash);
+        return key.data;
+    }
+
+    const std::vector<unsigned char> ToPKHBytes(const Key& key) const
+    {
+        if (key.is_hash) return key.data;
+        const auto h = Hash160(key.data);
+        return {h.begin(), h.end()};
+    }
+
+    template<typename I>
+    std::optional<Key> FromPKBytes(I first, I last) const
+    {
+        Key key;
+        key.data.assign(first, last);
+        key.is_hash = false;
+        return key;
+    }
+
+    template<typename I>
+    std::optional<Key> FromPKHBytes(I first, I last) const
+    {
+        Key key;
+        key.data.assign(first, last);
+        key.is_hash = true;
+        return key;
+    }
+} SCRIPT_PARSER_CONTEXT;
+
+} // namespace
+
+void FuzzInit()
+{
+    ECC_Start();
+    TEST_DATA.Init();
+}
+
+/* Fuzz tests that test parsing from a string, and roundtripping via string. */
+FUZZ_TARGET_INIT(miniscript_string, FuzzInit)
+{
+    FuzzedDataProvider provider(buffer.data(), buffer.size());
+    auto str = provider.ConsumeRemainingBytesAsString();
+    auto parsed = miniscript::FromString(str, PARSER_CTX);
```

Comment:
Looks like this may take quite a while. For example:

[clusterfuzz-testcase-miniscript_string-5717444439703552.bin.txt](https://github.com/bitcoin/bitcoin/files/8867405/clusterfuzz-testcase-miniscript_string-5717444439703552.bin.txt)

takes 10 seconds

With the flame graph:

![Screenshot from 2022-06-09 08-14-22](https://user-images.githubusercontent.com/6399679/172777704-a2110ad7-dbb9-4469-a94b-08767745db32.png)

Reply:
Yeah calling it in the constructor was not exactly bright since we'll re-do the same computation as we pile more nodes in the parsers. On the other hand it's hard not to do so and to keep the `const`ness of the `NodeRef`. Here are different approaches i tried:
- Cache the duplicate key result only once we call it https://github.com/darosior/bitcoin/tree/miniscript_cache_dup_key_check. Needs to drop the constness.
- Conditionally skip the check in the constructor, in order to only do it for the top level node https://github.com/darosior/bitcoin/tree/miniscript_skip_dup_key_check. Doesn't drop the constness but not elegant..
- Don't cache it at all https://github.com/darosior/bitcoin/tree/miniscript_no_dup_key_cache

Reply:
@darosior What about this idea:

* Make the current `miniscript::Node` constructors private, and make them not compute the `duplicate_key` value.
* Add a forwarding public constructor which invokes the private ones, and also computes the `duplicate_key` value (so now any publicly constructed Node object will have duplicate_key set, but it's possible to bypass this computation by using the private ones, if you have access to them).
* `FromScript` and `FromString` are made `friend`, and thus can use the private constructors.
* Make the `duplicate_key` variable mutable, so `ContainsDuplicateKey` can modify it (and is changed so it sets the value for all children of the called node too).

Reply:
Sounds much better than all my approaches. I'll implement and PR this.

Reply:
See also https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=47994